it/trace-context-leak/src/test/java/com/linecorp/armeria/internal/common/TraceRequestContextLeakTest.java

```diff
@@ -0,0 +1,289 @@
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.testing.junit5.common.EventLoopExtension;
+import com.linecorp.armeria.testing.junit5.common.EventLoopGroupExtension;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+
+class TraceRequestContextLeakTest {
+
+    @RegisterExtension
+    static final EventLoopExtension eventLoopExtension = new EventLoopExtension();
+
+    @RegisterExtension
+    static final EventLoopGroupExtension eventLoopGroupExtension = new EventLoopGroupExtension(2);
+
+    @Test
+    void singleThreadContextNotLeak() throws InterruptedException {
+        final AtomicBoolean isThrown = new AtomicBoolean(false);
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        final EventLoop executor =  eventLoopExtension.get();
+
+        executor.execute(() -> {
+            final ServiceRequestContext ctx = newCtx("/1");
+            try (SafeCloseable ignore = ctx.push()) {
+                //Ignore
+            } catch (Exception ex) {
+                isThrown.set(true);
+            } finally {
+                latch.countDown();
+            }
+        });
+
+        executor.execute(() -> {
+            final ServiceRequestContext anotherCtx = newCtx("/2");
+            try (SafeCloseable ignore = anotherCtx.push()) {
+                final ClientRequestContext clientCtx = newClientCtx("/3");
+                try (SafeCloseable ignore2 = clientCtx.push()) {
+                    //Ignore
+                }
+            } catch (Exception ex) {
+                isThrown.set(true);
+            } finally {
+                latch.countDown();
+            }
+        });
+
+        latch.await();
+        assertThat(isThrown).isFalse();
+    }
+
+    @Test
+    @SuppressWarnings("MustBeClosedChecker")
+    void singleThreadContextLeak() throws InterruptedException {
+        final AtomicBoolean isThrown = new AtomicBoolean();
+        final AtomicReference<Exception> exception = new AtomicReference<>();
+
+        try (DeferredClose deferredClose = new DeferredClose()) {
+            final EventLoop executor =  eventLoopExtension.get();
+
+            executor.execute(() -> {
+                final ServiceRequestContext ctx = newCtx("/1");
+                final SafeCloseable leaked = ctx.push(); // <- Leaked, should show in error.
+                deferredClose.add(executor, leaked);
+            });
+
+            executor.execute(() -> {
+                final ServiceRequestContext anotherCtx = newCtx("/2");
+                try (SafeCloseable ignore = anotherCtx.push()) {
+                    //Ignore
+                } catch (Exception ex) {
+                    isThrown.set(true);
+                    exception.set(ex);
```

Comment:
For correct happen-before relationship. 
```suggestion
                    exception.set(ex);
                    isThrown.set(true);
```

Reply:
Oh! Thanks for pointing out that.🙏